File: framework/report.py
from __future__ import annotations

import json
import shutil
from dataclasses import asdict
from pathlib import Path
from typing import Dict, List

from .utils import canonical_json, ensure_dir


def _junit_xml(run: Dict) -> str:
    tests = len(run["steps"])
    failures = sum(1 for s in run["steps"] if s["status"] != "ok")
    lines: List[str] = []
    lines.append(f'<testsuite name="{run["run_id"]}" tests="{tests}" failures="{failures}">')
    for s in run["steps"]:
        lines.append(f'  <testcase name="{s["id"]}" classname="visor.{run["platform"]}" time="{s["duration_ms"] / 1000:.3f}">')
        if s["status"] != "ok":
            msg = (s.get("error") or {}).get("message", "step failed")
            lines.append(f"    <failure message=\"{msg}\" />")
        lines.append("  </testcase>")
    lines.append("</testsuite>")
    return "\n".join(lines) + "\n"


def _materialize_artifacts(root: Path, artifact_paths: List[str]) -> List[str]:
    screenshots_dir = ensure_dir(root / "screenshots")
    sources_dir = ensure_dir(root / "sources")
    persisted: List[str] = []
    for i, artifact in enumerate(artifact_paths, start=1):
        src = Path(artifact)
        if not src.exists():
            continue
        base_dir = screenshots_dir if src.suffix.lower() == ".png" else sources_dir
        dest = base_dir / f"{i:03d}-{src.name}"
        if src.resolve() != dest.resolve():
            shutil.copy2(src, dest)
        persisted.append(str(dest))
    return persisted


def write_reports(result, report_dir: str) -> Dict[str, str]:
    root = ensure_dir(Path(report_dir) / result.run_id)
    env = ensure_dir(root / "env")

    payload = asdict(result)
    persisted_artifacts = _materialize_artifacts(root, result.artifacts)
    payload["artifacts"] = persisted_artifacts
    result.artifacts = persisted_artifacts

    summary_txt = root / "summary.txt"
    summary_json = root / "summary.json"
    junit_xml = root / "junit.xml"
    timeline_log = root / "timeline.log"

    summary_txt.write_text(
        "\n".join(
            [
                f"run_id: {result.run_id}",
                f"platform: {result.platform}",
                f"device: {result.device}",
                f"status: {result.status.value}",
                f"determinism_signature: {result.determinism_signature}",
                f"artifact_count: {len(result.artifacts)}",
            ]
        )
        + "\n",
        encoding="utf-8",
    )

    summary_json.write_text(canonical_json(payload) + "\n", encoding="utf-8")
    junit_xml.write_text(_junit_xml(payload), encoding="utf-8")

    timeline_lines = [f"{i + 1:03d} {s.id} {s.command} {s.status.value} {s.duration_ms}ms" for i, s in enumerate(result.steps)]
    timeline_log.write_text("\n".join(timeline_lines) + "\n", encoding="utf-8")

    (env / "runtime.json").write_text(
        json.dumps({"seed": result.seed, "started_at": result.started_at, "ended_at": result.ended_at}, indent=2) + "\n",
        encoding="utf-8",
    )
    (env / "device.json").write_text(json.dumps({"platform": result.platform, "device": result.device}, indent=2) + "\n", encoding="utf-8")

    report_html = root / "report.html"
    report_html.write_text(
        f"<html><body><h1>{result.run_id}</h1><p>Status: {result.status.value}</p><p>Artifacts: {len(result.artifacts)}</p></body></html>\n",
        encoding="utf-8",
    )

    return {
        "summary": str(summary_txt),
        "json": str(summary_json),
        "junit": str(junit_xml),
        "timeline": str(timeline_log),
        "html": str(report_html),
    }

File: src/report.ts
import fs from 'node:fs';
import path from 'node:path';

import type { RunResult } from './types.js';
import { canonicalJson, ensureDir } from './utils.js';

function escapeXml(value: string): string {
  return value
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
    .replaceAll("'", '&apos;');
}

function junitXml(run: RunResult): string {
  const tests = run.steps.length;
  const failures = run.steps.filter((step) => step.status !== 'ok').length;
  const lines: string[] = [
    `<testsuite name="${escapeXml(run.run_id)}" tests="${tests}" failures="${failures}">`
  ];

  for (const step of run.steps) {
    lines.push(
      `  <testcase name="${escapeXml(step.id)}" classname="visor.${escapeXml(run.platform)}" time="${(
        step.duration_ms / 1000
      ).toFixed(3)}">`
    );
    if (step.status !== 'ok') {
      const message = step.error?.message ?? 'step failed';
      lines.push(`    <failure message="${escapeXml(message)}" />`);
    }
    lines.push('  </testcase>');
  }

  lines.push('</testsuite>');
  return `${lines.join('\n')}\n`;
}

function materializeArtifacts(root: string, artifactPaths: string[]): string[] {
  const screenshotsDir = ensureDir(path.join(root, 'screenshots'));
  const sourcesDir = ensureDir(path.join(root, 'sources'));
  const persisted: string[] = [];

  artifactPaths.forEach((artifactPath, index) => {
    if (!fs.existsSync(artifactPath)) {
      return;
    }

    const sourcePath = path.resolve(artifactPath);
    const destinationRoot = sourcePath.toLowerCase().endsWith('.png') ? screenshotsDir : sourcesDir;
    const destinationPath = path.join(destinationRoot, `${String(index + 1).padStart(3, '0')}-${path.basename(sourcePath)}`);

    if (sourcePath !== path.resolve(destinationPath)) {
      fs.copyFileSync(sourcePath, destinationPath);
    }

    persisted.push(destinationPath);
  });

  return persisted;
}

export function writeReports(result: RunResult, reportDir: string): Record<string, string> {
  const root = ensureDir(path.join(reportDir, result.run_id));
  const env = ensureDir(path.join(root, 'env'));
  const payload = structuredClone(result);
  const persistedArtifacts = materializeArtifacts(root, result.artifacts);

  payload.artifacts = persistedArtifacts;
  result.artifacts = persistedArtifacts;

  const summaryTxt = path.join(root, 'summary.txt');
  const summaryJson = path.join(root, 'summary.json');
  const junitPath = path.join(root, 'junit.xml');
  const timelinePath = path.join(root, 'timeline.log');
  const htmlPath = path.join(root, 'report.html');

  fs.writeFileSync(
    summaryTxt,
    [
      `run_id: ${result.run_id}`,
      `platform: ${result.platform}`,
      `device: ${result.device}`,
      `status: ${result.status}`,
      `determinism_signature: ${result.determinism_signature}`,
      `artifact_count: ${result.artifacts.length}`,
      ''
    ].join('\n'),
    'utf8'
  );

  fs.writeFileSync(summaryJson, `${canonicalJson(payload)}\n`, 'utf8');
  fs.writeFileSync(junitPath, junitXml(payload), 'utf8');
  fs.writeFileSync(
    timelinePath,
    `${payload.steps
      .map((step, index) => `${String(index + 1).padStart(3, '0')} ${step.id} ${step.command} ${step.status} ${step.duration_ms}ms`)
      .join('\n')}\n`,
    'utf8'
  );
  fs.writeFileSync(
    path.join(env, 'runtime.json'),
    `${JSON.stringify(
      {
        seed: result.seed,
        started_at: result.started_at,
        ended_at: result.ended_at
      },
      null,
      2
    )}\n`,
    'utf8'
  );
  fs.writeFileSync(
    path.join(env, 'device.json'),
    `${JSON.stringify(
      {
        platform: result.platform,
        device: result.device
      },
      null,
      2
    )}\n`,
    'utf8'
  );
  fs.writeFileSync(
    htmlPath,
    `<html><body><h1>${escapeXml(result.run_id)}</h1><p>Status: ${escapeXml(result.status)}</p><p>Artifacts: ${result.artifacts.length}</p></body></html>\n`,
    'utf8'
  );

  return {
    summary: summaryTxt,
    json: summaryJson,
    junit: junitPath,
    timeline: timelinePath,
    html: htmlPath
  };
}
